### ISSUE TITLE
`Frozen` module for transfer learning.

### ISSUE THREAD

**Issue by tillahoffmann** [pytorch/pytorch#83492]
### 🚀 The feature, motivation and pitch

"Freezing" a model to stop gradient propagation and parameter updates is a common step in transfer learning. The [tutorial on transfer learning](https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html#convnet-as-fixed-feature-extractor) currently suggests setting `requires_grad = False` for all parameters of the feature extractor and removing the parameters from the set of parameters passed to the optimizer. What would be your thoughts on adding a `Frozen` module that achieves the same task as a one-liner? Proposed implementation below.

```python
class Frozen(torch.nn.Module):
    def __init__(self, frozen_module):
        super().__init__()
        self.frozen_module = frozen_module

    def forward(self, *args, **kwargs):
        with torch.no_grad():
            return self.frozen_module(*args, **kwargs)

    def parameters(self, recurse: bool = True) -> typing.Iterator[th.nn.Parameter]:
        return
        yield

    def named_parameters(self, prefix: str = '', recurse: bool = True) \
            -> typing.Iterator[tuple[str, th.nn.Parameter]]:
        return
        yield
```

Usage example: 

```python
resnet18 = torchvision.models.resnet18(pretrained=True)
extractor = Frozen(resnet18)  # Stops all gradients and parameter lookups.
```

As a side-effect, this would allow gradients to propagate if there are multiple paths from loss to parameters, e.g., in multi-objective optimization. Calling `extractor` would not propagate gradients. Calling `resnet18` directly would propagate gradients.

**Comment by albanD** [comment]
Hi,

I think this is an interesting proposal but there are a few limitations with the proposed implementation.
In particular what if the current Module is in the middle of the network, this implementation will prevent gradients from flowing to the earlier part of the net.
Also you can actually remove the parameters and re-register them as plain Tensors instead of overriding these methods.
